internal/manifest/manifest.go

```diff
@@ -0,0 +1,194 @@
+	Arch    string `json:"arch,omitempty"`
+}
+
+type Slice struct {
+	Kind string `json:"kind"`
+	Name string `json:"name,omitempty"`
+}
+
+type Path struct {
+	Kind      string   `json:"kind"`
+	Path      string   `json:"path,omitempty"`
+	Mode      string   `json:"mode,omitempty"`
+	Slices    []string `json:"slices,omitempty"`
+	Hash      string   `json:"sha256,omitempty"`
+	FinalHash string   `json:"final_sha256,omitempty"`
+	Size      uint64   `json:"size,omitempty"`
+	Link      string   `json:"link,omitempty"`
+}
+
+type Content struct {
+	Kind  string `json:"kind"`
+	Slice string `json:"slice,omitempty"`
+	Path  string `json:"path,omitempty"`
+}
+
+type Manifest struct {
+	db *jsonwall.DB
+}
+
+// Read loads a Manifest from a file without performing any validation. The file
+// is assumed to be both valid jsonwall and a valid Manifest (see [Validate]).
+func Read(absPath string) (manifest *Manifest, err error) {
+	defer func() {
+		if err != nil {
+			err = fmt.Errorf("cannot read manifest: %s", err)
+		}
+	}()
+
+	file, err := os.OpenFile(absPath, os.O_RDONLY, 0644)
+	if err != nil {
+		return nil, err
+	}
+	defer file.Close()
+	r, err := zstd.NewReader(file)
+	if err != nil {
+		return nil, err
+	}
+	defer r.Close()
+	db, err := jsonwall.ReadDB(r)
+	if err != nil {
+		return nil, err
+	}
+	mfestSchema := db.Schema()
+	if mfestSchema != schema {
+		return nil, fmt.Errorf("unknown schema version %q", mfestSchema)
+	}
+
+	manifest = &Manifest{db: db}
+	return manifest, nil
+}
+
+func (manifest *Manifest) IteratePaths(pathPrefix string, onMatch func(Path) error) (err error) {
+	return iteratePrefix(manifest, Path{Kind: "path", Path: pathPrefix}, onMatch)
+}
+
+func (manifest *Manifest) IteratePackages(onMatch func(Package) error) (err error) {
+	return iteratePrefix(manifest, Package{Kind: "package"}, onMatch)
+}
+
+func (manifest *Manifest) IterateSlices(pkgName string, onMatch func(Slice) error) (err error) {
+	return iteratePrefix(manifest, Slice{Kind: "slice", Name: pkgName}, onMatch)
+}
+
+func (manifest *Manifest) IterateContents(slice string, onMatch func(Content) error) (err error) {
+	return iteratePrefix(manifest, Content{Kind: "content", Slice: slice}, onMatch)
+}
+
+// Validate checks that the Manifest is valid. Note that to do that it has to
+// load practically the whole manifest into memory and unmarshall all the
+// entries.
+func Validate(manifest *Manifest) (err error) {
+	defer func() {
+		if err != nil {
+			err = fmt.Errorf("invalid manifest: %s", err)
+		}
+	}()
+
+	pkgExist := map[string]bool{}
+	err = manifest.IteratePackages(func(pkg Package) error {
+		pkgExist[pkg.Name] = true
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+
+	sliceExist := map[string]bool{}
+	err = manifest.IterateSlices("", func(slice Slice) error {
+		sk, err := setup.ParseSliceKey(slice.Name)
+		if err != nil {
+			return err
+		}
+		if !pkgExist[sk.Package] {
+			return fmt.Errorf(`package %q not found in packages`, sk.Package)
```

Comment:
Per earlier comment, let's please use double quotes everywhere unless it actually makes sense for readability.  That's already the convention in the code, but even in this PR itself it's being used inconsistently, as we can see strings going either way without a reason.

Reply:
Done in https://github.com/canonical/chisel/pull/144/commits/56dd49ca33944ab4e09cc7778cdaec7245732924.